src/main/java/software/amazon/awssdk/crt/io/TlsKeyOperation.java

```diff
@@ -0,0 +1,226 @@
+        }
+
+        Type(int nativeValue) {
+            this.nativeValue = nativeValue;
+        }
+
+        public int getNativeValue() {
+            return nativeValue;
+        }
+
+        int nativeValue;
+        static Map<Integer, Type> enumMapping = buildEnumMapping();
+    }
+
+    private boolean closeCalled;
+    private boolean clearCalled;
+    private byte[] inputData;
+    private Type operationType;
+    private TlsSignatureAlgorithm signatureAlgorithm;
+    private TlsHashAlgorithm digestAlgorithm;
+
+    /* Called from native when there's a new operation to be performed. Creates a new TlsKeyOperation */
+    protected TlsKeyOperation(long nativeHandle, byte[] inputData, int operationType, int signatureAlgorithm,
+            int digestAlgorithm) {
+
+        acquireNativeHandle(nativeHandle);
+
+        this.closeCalled = false;
+        this.clearCalled = false;
+        this.inputData = inputData;
+        this.operationType = Type.getEnumValueFromInteger(operationType);
+        this.signatureAlgorithm = TlsSignatureAlgorithm.getEnumValueFromInteger(signatureAlgorithm);
+        this.digestAlgorithm = TlsHashAlgorithm.getEnumValueFromInteger(digestAlgorithm);
+    }
+
+    /**
+     * Returns the input data from native that needs to be operated on using the private key.
+     * You can determine the operation that needs to be performed on the data using the getType function.
+     *
+     * @return The input data from native that needs to be operated on
+     */
+    public byte[] getInput() {
+        return inputData;
+    }
+
+    /**
+     * Returns the operation that needs to be performed.
+     *
+     * @return The operation that needs to be performed.
+     */
+    public Type getType() {
+        return operationType;
+    }
+
+    /**
+     * Returns the TLS algorithm used in the signature.
+     *
+     * @return The TLS algorithm used in the signature
+     */
+    public TlsSignatureAlgorithm getSignatureAlgorithm() {
+        return signatureAlgorithm;
+    }
+
+    /**
+     * Returns the TLS Hash algorithm used in the digest.
+     *
+     * @return The TLS Hash algorithm used in the digest
+     */
+    public TlsHashAlgorithm getDigestAlgorithm() {
+        return digestAlgorithm;
+    }
+
+    /**
+     * The function to call when you have modified the input data using the private key and are ready to
+     * return it for use in the MQTT TLS Handshake.
+     *
+     * @param output The modified input data that has been modified by the custom key operation
+     */
+    public void complete(byte[] output) {
+        if (getNativeHandle() == 0 || this.clearCalled == true) {
+            Log.log(LogLevel.Error, LogSubject.CommonGeneral,
+                "No native handle set in TlsKeyOperation! Cannot complete operation");
+            return;
+        }
+
+        tlsKeyOperationComplete(getNativeHandle(), output);
+        clear();
```

Comment:
need SOME kind of synchronization somewhere
either in java, or C

Java has a cool `synchronize` keyword that makes this kind of thing simple